dask/base.py

```diff
@@ -1518,10 +1518,20 @@ def get_name_from_key(key) -> str:
         return key[0]
     if isinstance(key, str):
         return key
-    raise TypeError(f"Expected str or tuple[str, Hashable, ...]; got {key}")
+    raise TypeError(f"Expected str or a tuple starting with str; got {key!r}")
 
 
-def replace_name_in_key(key, rename: Mapping[str, str]):
+@overload
+def replace_name_in_key(key: str, rename: Mapping[str, str]) -> str:
+    ...
+
+
+@overload
+def replace_name_in_key(key: Key, rename: Mapping[str, str]) -> Key:
+    ...
+
+
+def replace_name_in_key(key, rename):
```

Comment:
In pandas we would annotate this with the union, not sure if dask has a procedure here

Reply:
The reason I didn't use the union is that the input determines the output and you really don't want to use unions as a return type (https://github.com/python/mypy/issues/1693). I've replaced the `@overload` with a more compact constrained `TypeVar` that achieves the same.